File: 03_Combinatorics_Search/40_Backtracking/base_code.py
def solve_n_queens(n):
    """
    N-Queen 문제의 모든 해답을 찾는 백트래킹 뼈대
    """
    result = []
    # board[i] = j 는 "i번째 행의 j번째 열에 퀸이 있다"는 의미입니다.
    board = [-1] * n

    def is_valid(row, col):
        """
        현재 위치(row, col)에 퀸을 놓을 수 있는지 검사합니다. (가지치기 조건)
        """
        for i in range(row):
            # 1. 같은 열에 퀸이 있는지 확인 (board[i] == col)
            # 2. 대각선에 퀸이 있는지 확인 (행의 차이 == 열의 차이)
            if board[i] == col or abs(board[i] - col) == abs(i - row):
                return False
        return True

    def backtrack(row):
        # 1. 종료 조건: N개의 행에 모두 퀸을 무사히 배치했다면 정답 처리
        if row == n:
            result.append(board[:])
            return

        # 2. 탐색: 현재 행(row)의 0번 열부터 N-1번 열까지 퀸을 놓아봄
        for col in range(n):
            if is_valid(row, col): # ✨ 핵심: 유망한 자리인지(가지치기) 확인!
                board[row] = col   # 퀸을 놓음 (방문 처리)
                
                backtrack(row + 1) # 다음 행으로 넘어감 (재귀)
                
                board[row] = -1    # 퀸을 다시 뺌 (상태 복구 / 백트래킹)

    # 0번째 행부터 탐색 시작
    backtrack(0)
    return result

# 사용 예시
if __name__ == "__main__":
    n = 4
    solutions = solve_n_queens(n)
    
    print(f"{n}x{n} 체스판의 N-Queen 해답 개수: {len(solutions)}개")
    for idx, sol in enumerate(solutions, 1):
        print(f"해답 {idx}: {sol} (인덱스는 행, 값은 열을 의미)")

File: 03_Combinatorics_Search/40_Backtracking/BaseCode.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class BaseCode {

    public static List<int[]> solveNQueens(int n) {
        List<int[]> result = new ArrayList<>();
        int[] board = new int[n]; 
        
        backtrack(result, board, 0, n);
        return result;
    }

    /**
     * @param row 현재 퀸을 놓으려고 시도하는 행
     */
    private static void backtrack(List<int[]> result, int[] board, int row, int n) {
        // 1. 종료 조건: N개의 퀸을 모두 배치 완료
        if (row == n) {
            result.add(board.clone()); // 배열 복사본 저장
            return;
        }

        // 2. 탐색: 현재 행(row)의 0열부터 n-1열까지 시도
        for (int col = 0; col < n; col++) {
            if (isValid(board, row, col)) { // ✨ 가지치기 (Pruning)
                board[row] = col;           // 퀸 배치
                
                backtrack(result, board, row + 1, n); // 다음 행으로 이동
                
                // 자바는 값을 덮어쓰는 구조라 명시적인 복구 코드(board[row] = 0)가 없어도
                // 다음 for문 루프에서 덮어씌워지며 자연스럽게 백트래킹이 됩니다.
            }
        }
    }

    /**
     * (row, col)에 퀸을 놓을 수 있는지 확인합니다.
     */
    private static boolean isValid(int[] board, int row, int col) {
        // 현재 행(row) 이전까지 배치된 퀸들과 충돌하는지 검사
        for (int i = 0; i < row; i++) {
            // 같은 열에 있거나 || 대각선에 있는 경우 (행의 차이 == 열의 차이)
            if (board[i] == col || Math.abs(board[i] - col) == Math.abs(i - row)) {
                return false;
            }
        }
        return true;
    }

    // 사용 예시
    public static void main(String[] args) {
        int n = 4;
        List<int[]> solutions = solveNQueens(n);

        System.out.println(n + "x" + n + " 체스판의 N-Queen 해답 개수: " + solutions.size() + "개");
        for (int i = 0; i < solutions.size(); i++) {
            System.out.println("해답 " + (i + 1) + ": " + Arrays.toString(solutions.get(i)));
        }
    }
}